### ISSUE TITLE
Did you compare AP and FPS (rather than BFLOPS) of models ?

### ISSUE THREAD

**Issue by AlexeyAB** [WongKinYiu/CrossStagePartialNetworks#1]
@WongKinYiu Hi,

Did you compare `AP` and `1080ti fps` (rather than BFLOPS) of models ?

* CSPResNeXt50-PANet-SPP-GIoU https://github.com/WongKinYiu/CrossStagePartialNetworks#gpu-real-time-models

* CSPDarkNet-53-SPP-GIoU-Yolov3 that is based on `CSPDarkNet-53` https://github.com/WongKinYiu/CrossStagePartialNetworks#big-models

* default Yolov3-SPP model

With the same: network resolution, mini_batch=batch/subdivisions, data augmentation - to compare apples with apples.

**Comment by AlexeyAB** [comment]
I found: https://arxiv.org/abs/1911.11929v1

* CSPResNeXt50-PANet-SPP-GIoU - 35 FPS (**+5 FPS**) - 60.6 mAP@0.5 - 38.4 mAP@0.5...0.95 (**+2.2**)

* Yolov3-SPP-(GIoU ???) - 30 FPS - 60.6 mAP@0.5 - 36.2 mAP@0.5...0.95

But did you try backbones `CSPDarkNet-53-SPP` and `CSPDarkNet-53-Elastic` for Yolov3-spp / PAN?

![image](https://user-images.githubusercontent.com/4096485/69805791-7cb1b400-11f2-11ea-8576-0b0942c4e780.png)

**Comment by WongKinYiu** [comment]
@AlexeyAB Hello,

Yes, we do comparison of AP, fps, BFLOPs, #parameters.
![image](https://user-images.githubusercontent.com/12152972/69805844-8cde8900-121c-11ea-9261-ff76ce244487.png)

For big models, due to I have only few GPUs.
Training one model need about one and half month.
So currently we only have results of **CSPResNeXt50-PANet-SPP** and **CSPResNeXt50-PANet-SPP-GIoU**.
If you would like to test the speed of other models, I could upload the cfg files.

**Comment by WongKinYiu** [comment]
> * CSPResNeXt50-PANet-SPP-GIoU - 35 FPS (**+5 FPS**) - 60.6 mAP@0.5 - 38.4 mAP@0.5...0.95 (**+2.2**)

This result is for **CSPResNeXt50-PANet-SPP**, not **CSPResNeXt50-PANet-SPP-GIoU**.
We focus on AP50, so in our paper, we do not use GIoU.
![image](https://user-images.githubusercontent.com/12152972/69806395-b5b34e00-121d-11ea-97d1-75af2ec16048.png)

**Comment by WongKinYiu** [comment]
@AlexeyAB Hello,

I upload the cfg files to [in progress](https://github.com/WongKinYiu/CrossStagePartialNetworks/tree/master/in%20progress).
Please fell free to try them.

**Comment by AlexeyAB** [comment]
@WongKinYiu Thanks!

Will you release cfg/weights files for these models? 

![image](https://user-images.githubusercontent.com/4096485/69807083-7244e980-11f5-11ea-8db9-288089bc70d8.png)

**Comment by WongKinYiu** [comment]
@AlexeyAB Hmm...

Due to the Non-disclosure agreement, I can not release the backbones of lightweight models currently.
If you are interested in the EFM(SAM) or PRN, I can share you the head cfg of these models.

**Comment by Code-Fight** [comment]
Hi @WongKinYiu 

Now it ’s "Next Month".

LOL

Waiting for you share.

**Comment by WongKinYiu** [comment]
@Code-Fight 
Tomorrow i ll discuss with them.

**Comment by Code-Fight** [comment]
@WongKinYiu Great
Looking forword your good news.

**Comment by WongKinYiu** [comment]
falid :sob: :sob: :sob:

**Comment by Code-Fight** [comment]
oh....too sad,
but thanks for your best

**Comment by dreambit** [comment]
@WongKinYiu Hi

I would like to test _**csresnet50-elastic-panet-spp.cfg, csresnext50-yolo-spp.cfg, csresnext50-panet-spp.cfg, csresnext50-panet-spp-giou.cfg**_ on gtx1080ti

What should i change in config files for classes = 2?

Is it enough to:
1) change line classes=80 to 2 in each of 3 [yolo]-layers
2) change [filters=255] to filters=(2+ 5)x3 in the 3 [convolutional] before each [yolo] layer
as in @AlexeyAB documentation?

Thank you.

**Comment by WongKinYiu** [comment]
@dreambit hello,

plz use the pretrained weights from https://github.com/WongKinYiu/CrossStagePartialNetworks#big-models

and follow the https://github.com/AlexeyAB/darknet#how-to-train-tiny-yolo-to-detect-your-custom-objects to get initial weights like csresnext50.conv.80 and so on.

**Comment by dreambit** [comment]
@WongKinYiu Thanks, what about changes in config files with different classes and network size?
are there any changes that have to be made except those mentioned in @AlexeyAB docs?

**Comment by WongKinYiu** [comment]
@dreambit 

> change line classes=80 to 2 in each of 3 [yolo]-layers
> change [filters=255] to filters=(2+ 5)x3 in the 3 [convolutional] before each [yolo] layer
as in @AlexeyAB documentation?

it is enough.

**Comment by dreambit** [comment]
@WongKinYiu Hello

I have almost completed training csresnext50-panet-spp-giou.cfg.
Could u provide (if you have one) elastic version of csresnext50-panet-spp[-giou], so i can test it.

Thank you.

**Comment by WongKinYiu** [comment]
@dreambit 

https://github.com/WongKinYiu/CrossStagePartialNetworks/blob/master/in%20progress/csresnext50-elastic-panet-spp.cfg
here u r.

**Comment by dreambit** [comment]
@WongKinYiu, appreciate it 👍

**Comment by dreambit** [comment]
@WongKinYiu 
I've changed network size to 640x385, and classes = 2 filters = 21(in each conv. layer before yolo), random = 0

Bug i have 0 mAP.

What is wrong?

![image](https://user-images.githubusercontent.com/1479661/71012376-18528800-2100-11ea-8237-4125ca5df6e6.png)


Thanks.

**Comment by AlexeyAB** [comment]
@dreambit 

Use pre-trained weights.

Set 608x384

https://github.com/AlexeyAB/darknet#how-to-improve-object-detection

> increase network resolution in your .cfg-file (height=608, width=608 or **any value multiple of 32**) - it will increase precision

**Comment by dreambit** [comment]
@AlexeyAB 

It was a typo, size is  640x384

@WongKinYiu @AlexeyAB I used wrong initial weights, now it is ok.

**Comment by WongKinYiu** [comment]
@dreambit great!

**Comment by cmtsai** [comment]
I have changed network size to 608x608, classes = 40, and  filters = 135 (in each conv. layer before yolo), random = 1.
I have used csresnext50-panet-spp.cfg and csresnext50c.conv.80 to train  my own data set.
However, current avg loss = -nan and 0 mAP.
What is the problem?

![chart](https://user-images.githubusercontent.com/12856526/71171679-5505b700-2299-11ea-8765-440f996e045a.png)

**Comment by WongKinYiu** [comment]
@cmtsai Hello,

csresnext50-panet-spp.cfg should use csresnext50.conv.80 as pre-trained model, not csresnext50c.conv.80.

**Comment by cmtsai** [comment]
@WongKinYiu Hello,
3Q very much for your response!
How to obtain csresnext50.conv.80?

**Comment by cmtsai** [comment]
@WongKinYiu Hi,
I got it!
3Q very much!

**Comment by AlexeyAB** [comment]
@WongKinYiu Hi,

Did you try to use Maxout for CSP?

I.e. after route use maxpool_depth:
```
[route]
layers = 36, 21

[maxpool]
maxpool_depth=1
out_channels=256
stride=1
size=1
```


![image](https://user-images.githubusercontent.com/4096485/71427720-93053f80-26cc-11ea-927f-ad9237353421.png)

**Comment by WongKinYiu** [comment]
@AlexeyAB 

Hello, I used maxout in small model.
and I applied maxpool_depth before route.

**Comment by AlexeyAB** [comment]
@WongKinYiu Hi,

Why did you add `CenterNet resnet-102` instead of `CenterNet dla-34`  to your comparison table?

**Comment by WongKinYiu** [comment]
@AlexeyAB 

Hello, we test `CenterNet dla-34` on 1080ti, but it can not reach > 30 fps.
In my test, `CenterNet resnet-102` run faster than `CenterNet dla-34` almost twice times.
So I only compare with `CenterNet resnet-102`.

There are also many issues of inference speed of dla on CenterNet's github.
I have not found solution to solve the inference time problem.
If you face the problem and solve it, please tell me how to deal with it.

**Comment by AlexeyAB** [comment]
@WongKinYiu 

* Did you check FPS and AP (accuracy) by using test-dev MSCOCO evaluation server of their `CenterNet resnet-102 + deform_conv`, or did you check only FPS ?

* Did you test `CenterNet dla-34` on: ` single-scale` without `flip` with Pytorch 0.4.1, CUDA 9.0, and CUDNN 7.1?

* Did you ask him what could be the issue? https://github.com/xingyizhou/CenterNet/issues

So their results aren't reproducible on 1080 Ti (GP102-350-K1-A1) which is the mostly the same as their Titan Xp (GP102-450-A1)

https://arxiv.org/pdf/1904.07850v2.pdf

> i7-8086K CPU, Titan Xp GPU, Pytorch 0.4.1, CUDA 9.0, and CUDNN 7.1.

**Comment by WongKinYiu** [comment]
@AlexeyAB 

I only check FPS.
I do not open an issue, however, in their discussion:
* windows will gets lower fps
* use cudnn will gets lower fps
* different pytorch version will gets lower fps (maybe modify dcn can solve it)
...

**Comment by AlexeyAB** [comment]
@WongKinYiu Typo in chart.

Initially theses 2 models was for 416x416 (53 FPS) and 512x512 (44 FPS)
but 2 new models are for 512x512 (44 FPS) and 608x608 (35 FPS):
![image](https://user-images.githubusercontent.com/4096485/71692884-31c02900-2dbc-11ea-9b3c-0a8259b83574.png)

**Comment by WongKinYiu** [comment]
@AlexeyAB Oh!

Thanks for reminder, i will correct the figure soon.

**Comment by WongKinYiu** [comment]
@AlexeyAB 

![image](https://user-images.githubusercontent.com/12152972/71701359-f21d2f80-2e03-11ea-9893-bb47196e19e7.png)

**Comment by WongKinYiu** [comment]
@AlexeyAB 

![image](https://user-images.githubusercontent.com/12152972/71701484-90a99080-2e04-11ea-9a5d-0d8f2ec4e33b.png)

**Comment by reactivetype** [comment]
> If you are interested in the EFM(SAM), I can share you the head cfg of these models.

@WongKinYiu  would you please share the part of cfg for EFM head. I would like to compare it with PRN and PANet

**Comment by WongKinYiu** [comment]
sure, will share the cfg for efm head few hours later.

**Comment by WongKinYiu** [comment]
```
##########################

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[shortcut]
activation=leaky
from=-2

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[upsample]
stride=2

# pyramid
[route]
layers= #13x13

[maxpool]
maxpool_depth=1
out_channels=128
stride=1
size=1

[route]
layers=-1,-3

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=128
size=3
stride=1
pad=1
activation=leaky

[shortcut]
activation=leaky
from=-2

[convolutional]
batch_normalize=1
filters=64
size=1
stride=1
pad=1
activation=leaky

# pyramid
[route]
layers= #26x26

[maxpool]
maxpool_depth=1
out_channels=64
stride=1
size=1

[convolutional]
batch_normalize=1
filters=64
size=3
stride=2
pad=1
activation=leaky

[route]
layers=-1,-4

[convolutional]
batch_normalize=1
filters=64
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=64
size=3
stride=1
pad=1
activation=leaky

[shortcut]
activation=leaky
from=-2

[convolutional]
batch_normalize=1
filters=32
size=1
stride=1
pad=1
activation=leaky

[upsample]
stride=2

# pyramid
[route]
layers= #26x26

[maxpool]
maxpool_depth=1
out_channels=32
stride=1
size=1

[route]
layers=-1,-3

[convolutional]
batch_normalize=1
filters=64
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=64
size=3
stride=1
pad=1
activation=leaky

[shortcut]
activation=leaky
from=-2

[convolutional]
batch_normalize=1
filters=32
size=1
stride=1
pad=1
activation=leaky

# pyramid
[route]
layers= #52x52

[maxpool]
maxpool_depth=1
out_channels=32
stride=1
size=1

[convolutional]
batch_normalize=1
size=3
stride=2
pad=1
filters=32
activation=leaky

[route]
layers=-1,-4

##########################
```

**Comment by ekko-1996** [comment]
> ```
> ##########################
> 
> [convolutional]
> batch_normalize=1
> filters=256
> size=1
> stride=1
> pad=1
> activation=leaky
> 
> [convolutional]
> batch_normalize=1
> filters=256
> size=3
> stride=1
> pad=1
> activation=leaky
> 
> [shortcut]
> activation=leaky
> from=-2
> 
> [convolutional]
> batch_normalize=1
> filters=128
> size=1
> stride=1
> pad=1
> activation=leaky
> 
> [upsample]
> stride=2
> 
> # pyramid
> [route]
> layers= #26x26
> 
> [maxpool]
> maxpool_depth=1
> out_channels=128
> stride=1
> size=1
> 
> [route]
> layers=-1,-3
> 
> [convolutional]
> batch_normalize=1
> filters=128
> size=1
> stride=1
> pad=1
> activation=leaky
> 
> [convolutional]
> batch_normalize=1
> filters=128
> size=3
> stride=1
> pad=1
> activation=leaky
> 
> [shortcut]
> activation=leaky
> from=-2
> 
> [convolutional]
> batch_normalize=1
> filters=64
> size=1
> stride=1
> pad=1
> activation=leaky
> 
> # pyramid
> [route]
> layers= #52x52
> 
> [maxpool]
> maxpool_depth=1
> out_channels=64
> stride=1
> size=1
> 
> [convolutional]
> batch_normalize=1
> filters=64
> size=3
> stride=2
> pad=1
> activation=leaky
> 
> [route]
> layers=-1,-4
> 
> [convolutional]
> batch_normalize=1
> filters=64
> size=1
> stride=1
> pad=1
> activation=leaky
> 
> [convolutional]
> batch_normalize=1
> filters=64
> size=3
> stride=1
> pad=1
> activation=leaky
> 
> [shortcut]
> activation=leaky
> from=-2
> 
> [convolutional]
> batch_normalize=1
> filters=32
> size=1
> stride=1
> pad=1
> activation=leaky
> 
> [upsample]
> stride=2
> 
> # pyramid
> [route]
> layers= #52x52
> 
> [maxpool]
> maxpool_depth=1
> out_channels=32
> stride=1
> size=1
> 
> [route]
> layers=-1,-3
> 
> [convolutional]
> batch_normalize=1
> filters=64
> size=1
> stride=1
> pad=1
> activation=leaky
> 
> [convolutional]
> batch_normalize=1
> filters=64
> size=3
> stride=1
> pad=1
> activation=leaky
> 
> [shortcut]
> activation=leaky
> from=-2
> 
> [convolutional]
> batch_normalize=1
> filters=32
> size=1
> stride=1
> pad=1
> activation=leaky
> 
> # pyramid
> [route]
> layers= #104x104
> 
> [maxpool]
> maxpool_depth=1
> out_channels=32
> stride=1
> size=1
> 
> [convolutional]
> batch_normalize=1
> size=3
> stride=2
> pad=1
> filters=32
> activation=leaky
> 
> [route]
> layers=-1,-4
> 
> ##########################
> ```

@WongKinYiu HI，I cant understand the cfg of EFM，I think it is different from the structure in the paper
